api/src/test/java/org/openmrs/obs/ImageHandlerTest.java

```diff
@@ -47,4 +77,87 @@ public void shouldNotSupportOtherViews() {
         assertFalse(handler.supportsView(""));
         assertFalse(handler.supportsView((String) null));
     }
+	
+	/** This method sets up the test data's parameters for the mime type tests  **/
+	@Before
+	public void initVariablesForMimetypeTests() {
+		filepath = new File("target" + File.separator + "test-classes").getAbsolutePath();
+		sourceFile = new File(
+		        "src" + File.separator + "test" + File.separator + "resources" + File.separator + "ComplexObsTestImage.png");
+	}
+	
+	@Test
+	public void shouldSaveMimeTypeInFilenamesAndReadIt() throws IOException {
+		final String mimetype = "image/png";
+		
+		BufferedImage img = ImageIO.read(sourceFile);
+		
+		ComplexData complexData = new ComplexData(FILENAME, img);
+		complexData.setMimeType(mimetype);
+		
+		// Construct 2 Obs to also cover the case where the filename exists already
+		Obs obs1 = new Obs();
+		obs1.setComplexData(complexData);
+		
+		Obs obs2 = new Obs();
+		obs2.setComplexData(complexData);
+		
+		// Mocked methods
+		mockStatic(Context.class);
+		when(Context.getAdministrationService()).thenReturn(administrationService);
+		when(administrationService.getGlobalProperty(any())).thenReturn(filepath);
+		
+		ImageHandler handler = new ImageHandler();
+		
+		// Execute save
+		handler.saveObs(obs1);
+		handler.saveObs(obs2);
+		
+		// Get observation
+		Obs complexObs = handler.getObs(obs1, "RAW_VIEW");
+		Obs complexObs2 = handler.getObs(obs2, "RAW_VIEW");
+		
+		assertTrue(complexObs.getComplexData().getMimeType().equals(mimetype));
+		assertTrue(complexObs2.getComplexData().getMimeType().equals(mimetype));
+		
+		// Delete created files to avoid cluttering
+		File obsFile1 = ImageHandler.getComplexDataFile(obs1);
+		File obsFile2 = ImageHandler.getComplexDataFile(obs2);
+		obsFile1.delete();
+		obsFile2.delete();
+	}
+	
+	@Test
+	public void shouldGuessMimeTypeIfNotSpecified() throws IOException {
+		final String mimetype = "image/png";
+
+		BufferedImage img = ImageIO.read(sourceFile);
+		
+		ComplexData complexData = new ComplexData(FILENAME, img);
+		complexData.setMimeType(null);
+		
+		Obs obs = new Obs();
+		obs.setComplexData(complexData);
+		
+		// Mocked methods
+		mockStatic(Context.class);
+		when(Context.getAdministrationService()).thenReturn(administrationService);
+		when(administrationService.getGlobalProperty(any())).thenReturn(filepath);
+		
+		ImageHandler handler = new ImageHandler();
+		
+		// Execute save
+		handler.saveObs(obs);
+		
+		// Get observation
+		Obs complexObs = handler.getObs(obs, "RAW_VIEW");
+		
+		String probedMimetype = complexObs.getComplexData().getMimeType();
+		
+		assertTrue(probedMimetype.equals(mimetype));
```

Comment:
When i run this test locally, probedMimeType equals application/octet-stream instead of image/png which the above line expects.

Reply:
That's interesting, for me it runs fine. That might be what those complaints I read about the inconsistent behaviour of probeContentType() across different OS was about. That's what I was talking about previously.

There would be 2 ways to handle this: 
1. I could let the old code for ImageHandler execute if an image is given application/octet-stream (I would do this for TextHandler analogously then) because its probing might be more reliable? I'd also adjust the test to accept the return mime type of the old code.
or
2 . Adjust the test to pass if either the image/png is recognized or application/octet-stream is returned because it could not be probed. 

Would you prefer one over the other?

Reply:
Having a test which expects inconsistent results, does not sound right.

Reply:
The problem is that probing (at least with Files.probeContentType()) is somehow system dependent and sometimes shows weird behaviour, at least from what I read. Probing is only a last resort if the mime type isn't in the filename, as it should be with all future complex obs. That is why I suggested option 1 in my previous post, in hopes that the old code returns more consistent types than Java's own probeContentType(). It's hard for me to evaluate though, with just 1 system here to test.

Another way to have consistency is to just determine the mime type (if it is not in the name) by looking at the file extension, which should be more consistent, but is incapable of guessing it from the file contents. That would be option 3, I think there is a method for that as well. That's why I ask what would be best for the project.

Thinking about it, having to guess the type from the file contents is probably not very important  and option 3 (just guessing it from the file extension) might be the best way.

Reply:
Did you see any disadvantage for using file extension?

Reply:
The only disadvantage is if no file extension exists either, we have to use a default type. Probing could try to find it out by reading the file itself instead of just the name + extension. But as you see, that's still kind of problematic.

I'm currently going for the file extension approach. The best way I have found is to modify META-INF/mime.types to contain more extensions and types, and then use MimetypesFileTypeMap. I will try this now and comment if I make any progress that I can push.